File: derapi_example.py
# derapi_example.py: Example client code to access Derapi services, in Python

import sys

import requests
import oauthlib.oauth2
import requests_oauthlib

class LoginInfo():
    """A simple class to hold login credentials for different services"""
    def __init__(self, token_url, client_id, client_secret):
        self.token_url = token_url
        self.client_id = client_id
        self.client_secret = client_secret

DERAPI_LOGIN = LoginInfo("https://auth.derapi.com/oauth2/token", "3tb5s5326rpd726gk8vcpllhj", "1nrvnklg39mlulik7u4q717q093q0htprng1p7teneg20vl6d020")

BACKEND_LOGINS = {
    "sma-sbox": LoginInfo("https://sandbox-auth.smaapis.de/oauth2/token", "derapi_api", "1q14ur3PxuiNlvlQGBzLrMPjyCzkCZAE"),
    "solis": LoginInfo("https://api.derapi.com/oauth/solis","1300386381676488475","e1596fd6a4f84e888327dbbc82ed8bd2"),
}

def token(token_url, client_id, client_secret):
    """Fetches an OAuth2 token from `token_url' using client credentials."""
    client = oauthlib.oauth2.BackendApplicationClient(client_id=client_id)
    session = requests_oauthlib.OAuth2Session(client=client)
    token = session.fetch_token(
        token_url=token_url,
        client_id=client_id,
        client_secret=client_secret)
    return token

def tokens(derapi_login, backend_logins):
    derapi_token = token(derapi_login.token_url,
                         derapi_login.client_id,
                         derapi_login.client_secret)
    backend_tokens = {}
    for backend_name in backend_logins:
        backend_tokens[backend_name] = token(backend_logins[backend_name].token_url,
                                             backend_logins[backend_name].client_id,
                                             backend_logins[backend_name].client_secret)
    return (derapi_token, backend_tokens)

def derapi_request(url, derapi_token, backend_tokens):
    x_auth_headers = {}
    for backend_name in backend_tokens:
        x_auth_headers[f"X-Authorization-{backend_name}"] = backend_tokens[backend_name]["access_token"]
    request = requests.Request("GET",
                               url=url,
                               auth=requests_oauthlib.OAuth2(token=derapi_token),
                               headers=x_auth_headers)
    return request.prepare()

def main():
    derapi_token, backend_tokens = tokens(DERAPI_LOGIN, BACKEND_LOGINS)
    request = derapi_request("https://api.derapi.com/sites", derapi_token, backend_tokens)
    response = requests.Session().send(request)
    print(response.text)

if __name__ == "__main__":
    sys.exit(main())

File: derapi_example.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

public class DerapiExample {

    public static String fetchToken(String tokenUrl, String clientId, String clientSecret) throws Exception {
        // Fetches an OAuth2 token from `token_url' using client credentials,
        // using standard Requests library facilities. An example token may
        // look like this:
        // {
        //   "access_token":"eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA",
        //   "expires_in":300,
        //   "refresh_expires_in":1800,
        //   "refresh_token":"eyJhbGcIgOiAiSldUIiwia2lkIiA6ICJhNmJlZjg4NS0yNT",
        // }
        String credentials = clientId + ":" + clientSecret;
        String encodedCredentials = Base64.getEncoder().encodeToString(credentials.getBytes());
        URL url = new URL(tokenUrl);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Authorization", "Basic " + encodedCredentials);
        connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
        connection.setDoOutput(true);
        String data = "grant_type=client_credentials";
        connection.getOutputStream().write(data.getBytes());

        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        String jsonResponse = response.toString();
        int start = jsonResponse.indexOf("\"access_token\":\"") + "\"access_token\":\"".length();
        int end = jsonResponse.indexOf("\"", start);
        return jsonResponse.substring(start, end);
    }


    public static Map tokens(LoginInfo derapiLogin, Map<String, LoginInfo> backendLogins) throws Exception {
        // Obtain OAuth2 tokens for Derapi and backends
        String derapiToken = fetchToken(derapiLogin.tokenUrl, derapiLogin.clientId, derapiLogin.clientSecret);

        Map backendTokens = new HashMap();

        for (Entry<String, LoginInfo> backend : backendLogins.entrySet()) {
            String backendToken = fetchToken(backend.getValue().tokenUrl, backend.getValue().clientId, backend.getValue().clientSecret);
            backendTokens.put(backend.getKey(), backendToken);
        }
        Map tokens = new HashMap();
        tokens.put("derapiToken", derapiToken);
        tokens.put("backendTokens", backendTokens);
        return tokens;
    }
    public static HttpURLConnection derapiRequest(String url, String derapiToken, Map<String, String> backendTokens) throws Exception {
    	// headers with Authorization header for Derapi and X-Authorization-*
    	// headers for backends. You need to do this for every Derapi URL you
    	// want to get. The headers may look like this:
    	// GET /sites HTTP/1.1
    	// Host: api.derapi.com
    	// Authorization: Bearer ZGVyYXBpIHRva2VuCg==
    	// X-Authorization-sma-sbox: Bearer eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA
    	// X-Authorization-solis: Bearer c29saXMgdG9rZW4K
    	Map<String, String> xAuthHeaders = new HashMap<>();
        for (String backendName : backendTokens.keySet()) {
            String accessToken = backendTokens.get(backendName);
            xAuthHeaders.put("X-Authorization-" + backendName, "Bearer " + accessToken);
        }

        URL requestUrl = new URL(url);
        HttpURLConnection connection = (HttpURLConnection) requestUrl.openConnection();
        connection.setRequestMethod("GET");
        connection.setRequestProperty("Authorization", "Bearer " + derapiToken);
        for (String headerName : xAuthHeaders.keySet()) {
            connection.setRequestProperty(headerName, xAuthHeaders.get(headerName));
        }
        return connection;
    }
    public static void main(String[] args) throws Exception {
    	// obtain OAuth2 tokens for Derapi, backends
        LoginInfo derapiLogin = new LoginInfo("https://auth.derapi.com/oauth2/token", "3tb5s5326rpd726gk8vcpllhj", "1nrvnklg39mlulik7u4q717q093q0htprng1p7teneg20vl6d020");
        Map backendLogins = new HashMap();

        LoginInfo smaLogin = new LoginInfo("https://sandbox-auth.smaapis.de/oauth2/token", "derapi_api", "1q14ur3PxuiNlvlQGBzLrMPjyCzkCZAE");
        LoginInfo solisLogin = new LoginInfo("https://api.derapi.com/oauth/solis", "1300386381676488475", "e1596fd6a4f84e888327dbbc82ed8bd2");
        backendLogins.put("sma-sbox", smaLogin);
        backendLogins.put("solis", solisLogin);
        // construct request for a Derapi resource using OAuth2 tokens
        Map<String, Object> tokens = tokens(derapiLogin, backendLogins);
        String derapiToken = (String) tokens.get("derapiToken");
        Map<String, String> backendTokens = (Map<String, String>) tokens.get("backendTokens");
        HttpURLConnection response = derapiRequest("https://api.derapi.com/sites", derapiToken, backendTokens);
        System.out.println(response.getResponseCode());

    }
}



// Remember to add this class in a different file.
public  class LoginInfo {
        public String tokenUrl;
        public String clientId;
        public String clientSecret;

        public LoginInfo(String tokenUrl, String clientId, String clientSecret) {
            this.tokenUrl = tokenUrl;
            this.clientId = clientId;
            this.clientSecret = clientSecret;
        }

 }
